### PR TITLE
planner: speed up PhysicalTopN#containVirtualColumn

### LINKED ISSUE #46809: Speed up PhysicalTopN#containVirtualColumn
## Enhancement
Currently `PhysicalTopN#containVirtualColumn` performs nested loop matching given columns with those of `ByItems`.

We can establish `map[int64]*expression.Column` at the beginning of the method keyed by column Ids of `tCols`.

This way, the loop over `p.ByItems` can utilize the `map` to perform O(1) lookup.

### PR DESCRIPTION
**Pull request by tedyu**
### What problem does this PR solve?

Issue Number: close #46809

Problem Summary:
Using nested loop is inefficient when column count is high.

### What is changed and how it works?
This is a refactoring of `PhysicalTopN#containVirtualColumn`.
At the beginning, we establish a `map` from column Id to `*expression.Column`.
In the main loop, we just check whether the underlying column Id is in the map.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by tedyu** [comment]
cc @hawkingrei

**Review by hawkingrei — APPROVED** [review]

File: planner/core/task.go [hunk 1]
```diff
@@ -1115,14 +1115,18 @@ func (p *PhysicalTopN) canExpressionConvertedToPB(storeTp kv.StoreType) bool {
 
 // containVirtualColumn checks whether TopN.ByItems contains virtual generated columns.
 func (p *PhysicalTopN) containVirtualColumn(tCols []*expression.Column) bool {
+	tColMap := make(map[int64]*expression.Column, len(tCols))
```
Comment (hawkingrei) [inline planner/core/task.go:1118]:
```suggestion
	tColSet := make(map[int64]struct{}, len(tCols))
```

It should be set. if it is map, the value is useless.

**Comment by tedyu** [comment]
```
make[1]: Leaving directory `/home/jenkins/agent/workspace/pingcap/tidb/ghpr_unit_test/tidb'
make: *** [check-bazel-prepare] Error 2
root: /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/execroot/__main__/bazel-out/k8-fastbuild/testlogs
mv: cannot stat 'bazel.xml': No such file or directory
script returned exit code 2
```
It seems the above error was not related to my PR.

**Comment by hawkingrei** [comment]
> ```
> make[1]: Leaving directory `/home/jenkins/agent/workspace/pingcap/tidb/ghpr_unit_test/tidb'
> make: *** [check-bazel-prepare] Error 2
> root: /home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/execroot/__main__/bazel-out/k8-fastbuild/testlogs
> mv: cannot stat 'bazel.xml': No such file or directory
> script returned exit code 2
> ```
> 
> It seems the above error was not related to my PR.

Yes, I think it is about our CI.

**Comment by tedyu** [comment]
```
 [ERROR] [runaway.go:145] ["try to get new runaway watch"] [error="[schema:1146]Table 'mysql.tidb_runaway_watch' doesn't exist"]
```
The above error doesn't seem to be related to my PR.

**Comment by hawkingrei** [comment]
> ```
>  [ERROR] [runaway.go:145] ["try to get new runaway watch"] [error="[schema:1146]Table 'mysql.tidb_runaway_watch' doesn't exist"]
> ```
> 
> The above error doesn't seem to be related to my PR.
> 
> I wonder if this may happen before all the upgrade scripts in `bootstrapVersion` complete execution. In this case, `upgradeToVer172` may not have run.

I found it. it is not your problem.

**Review by qw4990 — APPROVED** [review]

File: planner/core/task.go [hunk 2]
```diff
@@ -1115,14 +1115,18 @@ func (p *PhysicalTopN) canExpressionConvertedToPB(storeTp kv.StoreType) bool {
...
+		if tCol.ID > 0 && tCol.VirtualExpr != nil {
+			tColSet[tCol.ID] = struct{}{}
+		}
+	}
 	for _, by := range p.ByItems {
 		cols := expression.ExtractColumns(by.Expr)
 		for _, col := range cols {
-			for _, tCol := range tCols {
+			if _, ok := tColSet[col.ID]; ok {
 				// A column with ID > 0 indicates that the column can be resolved by data source.
```
Comment (hi-rustin) [inline planner/core/task.go:1128]: Could you please move this comment to the right place? Thanks!

**Review by hi-rustin — COMMENTED** [review]
👍🏿 

If we can add a golang benchmark to it, that would be nice.

File: planner/core/task.go (hunk 2 above)
Comment (tedyu) [reply planner/core/task.go:1128]: the PR has been merged.
If I happen to modify this class again, I can move it.

I think it is fine to leave the comment here as well.

**Pull request merged**